File: app/src/main/java/com/chebuso/chargetimer/charge/ChargeValuesProvider.java
package com.chebuso.chargetimer.charge;

import android.support.annotation.NonNull;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

public class ChargeValuesProvider {
    private final static int MAX_HOME_SOCKET_AMPERAGE = 16;

    public static List<String> getAllowedAmperage(int defaultAmperage){

        List<String> values = defaultAmperage > MAX_HOME_SOCKET_AMPERAGE
            ? getAllowedAmperageForPublicChargers(defaultAmperage)
            : getAllowedAmperageForHomeSockets(defaultAmperage);

        Collections.sort(values, new Comparator<String>() {
            @Override
            public int compare(String o1, String o2) {
                return Integer.parseInt(o1) - Integer.parseInt(o2);
            }
        });

        return values;
    }

    @NonNull
    private static List<String> getAllowedAmperageForHomeSockets(int defaultAmperage) {
        List<String> values;
        values = generateSequence(6, MAX_HOME_SOCKET_AMPERAGE, 2);

        if (!values.contains(String.valueOf(defaultAmperage)))
            values.add(String.valueOf(defaultAmperage));

        values.add("22");
        values.add("32");
        values.add("64");
        return values;
    }

    private static List<String> getAllowedAmperageForPublicChargers(int defaultAmperage) {
        List<String> values = generateSequence(8, MAX_HOME_SOCKET_AMPERAGE, 4);
        values.add(String.valueOf(defaultAmperage));

        final int MAX_STEPS = 8;

        int minAmperage = defaultAmperage / 2;
        if (minAmperage < MAX_HOME_SOCKET_AMPERAGE)
            minAmperage = MAX_HOME_SOCKET_AMPERAGE;

        final int maxAmperage = defaultAmperage * 2;
        final int step = (maxAmperage - minAmperage) / MAX_STEPS;

        List<String> additionalValues = generateSequence(minAmperage, maxAmperage, step);

        for (String tmpValue : additionalValues) {
            if (!values.contains(tmpValue))
                values.add(tmpValue);
        }
        return values;
    }

    public static List<String> getAllowedVoltage(int defaultVoltage){
        final int VOLTAGE_STEP = 5;

        int delta = getMinMaxVoltageDelta(defaultVoltage);
        if (delta > defaultVoltage) delta = defaultVoltage;

        final int min = defaultVoltage - delta;
        final int max = defaultVoltage + delta;

        return getAllowedVoltages(min, max, VOLTAGE_STEP);
    }

    private static int getMinMaxVoltageDelta(int defaultVoltage){
        final int MAX_US_VOLTAGE = 140;
        final int US_DELTA = 20;
        final int DEFAULT_DELTA = 40;

        final boolean isUSVoltage = defaultVoltage < MAX_US_VOLTAGE;
        return isUSVoltage ? US_DELTA : DEFAULT_DELTA;
    }

    private static List<String> getAllowedVoltages(int minVoltage, int maxVoltage, int step){
        return generateSequence(minVoltage, maxVoltage, step);
    }

    private static List<String> generateSequence(int min, int max, int step){
        List<String> allowedValues = new ArrayList<>();

        for (int i = min; i <= max; i += step){
            allowedValues.add(String.valueOf(i));
        }

        return allowedValues;
    }
}

File: EvChargeTimer/app/src/main/java/com/chebuso/chargetimer/charge/ChargeValuesProvider.kt
package com.chebuso.chargetimer.charge

object ChargeValuesProvider {
    private const val MAX_HOME_SOCKET_AMPERAGE = 16

    fun getAllowedAmperage(defaultAmperage: Int): List<String> {
        val values =
            if (defaultAmperage > MAX_HOME_SOCKET_AMPERAGE)
                getAllowedAmperageForPublicChargers(defaultAmperage)
            else
                getAllowedAmperageForHomeSockets(defaultAmperage)

        return values.sortedWith{ o1, o2 -> o1.toInt() - o2.toInt()}
    }

    private fun getAllowedAmperageForHomeSockets(defaultAmperage: Int): List<String> {
        val values: MutableList<String> = generateSequence(6, MAX_HOME_SOCKET_AMPERAGE, 2)
        if (!values.contains(defaultAmperage.toString()))
            values.add(defaultAmperage.toString())
        values.add("22")
        values.add("32")
        values.add("64")
        return values
    }

    private fun getAllowedAmperageForPublicChargers(defaultAmperage: Int): List<String> {
        val values = generateSequence(8, MAX_HOME_SOCKET_AMPERAGE, 4)
        values.add(defaultAmperage.toString())
        val MAX_STEPS = 8
        var minAmperage = defaultAmperage / 2
        if (minAmperage < MAX_HOME_SOCKET_AMPERAGE) minAmperage = MAX_HOME_SOCKET_AMPERAGE
        val maxAmperage = defaultAmperage * 2
        val step = (maxAmperage - minAmperage) / MAX_STEPS
        val additionalValues: List<String> = generateSequence(minAmperage, maxAmperage, step)
        for (tmpValue in additionalValues) {
            if (!values.contains(tmpValue))
                values.add(tmpValue)
        }
        return values
    }

    fun getAllowedVoltage(defaultVoltage: Int): List<String> {
        var delta = getMinMaxVoltageDelta(defaultVoltage)
        if (delta > defaultVoltage) delta = defaultVoltage
        val min = defaultVoltage - delta
        val max = defaultVoltage + delta
        return getAllowedVoltages(min, max)
    }

    private fun getMinMaxVoltageDelta(defaultVoltage: Int): Int {
        val MAX_US_VOLTAGE = 140
        val US_DELTA = 20
        val DEFAULT_DELTA = 40
        val isUSVoltage = defaultVoltage < MAX_US_VOLTAGE
        return if (isUSVoltage) US_DELTA else DEFAULT_DELTA
    }

    private fun getAllowedVoltages(minVoltage: Int, maxVoltage: Int): List<String> {
        val VOLTAGE_STEP = 5
        return generateSequence(minVoltage, maxVoltage, VOLTAGE_STEP)
    }

    private fun generateSequence(min: Int, max: Int, step: Int): MutableList<String> {
        val allowedValues: MutableList<String> = ArrayList()
        for(i in min..max step step){
            allowedValues.add(i.toString())
        }
        return allowedValues
    }
}

